javalib/src/main/scala/java/util/regex/PatternCompiler.scala

```diff
@@ -0,0 +1,1759 @@
+
+    for (i <- start until end) {
+      if (!isHexDigit(pattern.charAt(i)))
+        parseError(s"Illegal $nameForError escape sequence")
+    }
+
+    val cp =
+      if (end - start > 6) Character.MAX_CODE_POINT + 1
+      else parseInt(pattern.substring(start, end), 16)
+    if (cp > Character.MAX_CODE_POINT)
+      parseError("Hexadecimal codepoint is too big")
+
+    cp
+  }
+
+  /** Parses and returns a translated version of a pre-defined character class. */
+  private def parsePredefinedCharacterClass(dispatchChar: Char): CompiledCharClass = {
+    import CompiledCharClass._
+
+    pIndex += 1
+
+    val positive = (dispatchChar: @switch) match {
+      case 'd' | 'D' =>
+        if (unicodeCharacterClass) UnicodeDigit
+        else ASCIIDigit
+      case 'h' | 'H' =>
+        UniversalHorizontalWhiteSpace
+      case 's' | 'S' =>
+        if (unicodeCharacterClass) UnicodeWhitespace
+        else ASCIIWhiteSpace
+      case 'v' | 'V' =>
+        UniversalVerticalWhiteSpace
+      case 'w' | 'W' =>
+        wordCharClass
+      case 'p' | 'P' =>
+        parsePCharacterClass()
+    }
+
+    if (dispatchChar >= 'a') // cheap isLower
+      positive
+    else
+      positive.negated
+  }
+
+  /** Parses and returns a translated version of a `\p` character class. */
+  private def parsePCharacterClass(): CompiledCharClass = {
+    val pattern = this.pattern // local copy
+    val len = pattern.length()
+
+    val start = pIndex
+    val property = if (start == len) {
+      "?" // mimics the behavior of the JVM
+    } else if (pattern.charAt(start) == '{') {
+      val innerStart = start + 1
+      val innerEnd = pattern.indexOf("}", innerStart)
+      if (innerEnd < 0)
+        parseError("Unclosed character family")
+      pIndex = innerEnd
+      pattern.substring(innerStart, innerEnd)
+    } else {
+      pattern.substring(start, start + 1)
+    }
+
+    val result = if (!unicodeCharacterClass && asciiPOSIXCharacterClasses.contains(property)) {
+      val property2 =
+        if (asciiCaseInsensitive && (property == "Lower" || property == "Upper")) "Alpha"
+        else property
+      asciiPOSIXCharacterClasses(property2)
+    } else {
+      // For anything else, we need built-in support for \p
+      requireES2018Features("Unicode character family")
+
+      predefinedPCharacterClasses.getOrElse(property, {
+        val scriptPrefixLen = if (property.startsWith("Is")) {
+          2
+        } else if (property.startsWith("sc=")) {
+          3
+        } else if (property.startsWith("script=")) {
+          7
+        } else if (property.startsWith("In") || property.startsWith("blk=") || property.startsWith("block=")) {
+          parseError("Blocks are not supported in \\p Unicode character families")
+        } else {
+          // Error
+          parseError(s"Unknown Unicode character class '$property'")
+        }
+        CompiledCharClass.posP("sc=" + canonicalizeScriptName(property.substring(scriptPrefixLen)))
```

Comment:
Tying this back to your rebuttal in: https://github.com/scala-js/scala-js/pull/4455#discussion_r651064387 (which I agree with in principle). What will happen if somebody gives an invalid script name? (e.g. `\p{sc=frobber}`)?